src/dotty/tools/dotc/parsing/Parsers.scala

```diff
@@ -1308,7 +1319,28 @@ object Parsers {
      */
     val pattern2 = () => infixPattern() match {
       case p @ Ident(name) if isVarPattern(p) && in.token == AT =>
-        atPos(p.pos.start, in.skipToken()) { Bind(name, infixPattern()) }
+        val pos = in.skipToken()
+
+        // compatibility for Scala2 `x @ _*` syntax
+        infixPattern() match {
+          case pt @ Ident(tpnme.WILDCARD_STAR) =>
```

Comment:
3. Then this code can be written shorter like this:

        migrationWarningOrError("The syntax `x @ _*' is no longer supported; use `x : _*' instead", p.pos)